core/src/main/java/com/graphhopper/util/EngineWarmUp.java

```diff
@@ -1,26 +1,51 @@
 package com.graphhopper.util;
 
+import com.graphhopper.GHRequest;
+import com.graphhopper.GraphHopper;
 import com.graphhopper.storage.GraphHopperStorage;
-import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
 
 public class EngineWarmUp {
     /**
      * 'Warm up' virtual machine so initial route calculation is not slower than consequent. To be used
      * on a less powerful machine like Android or Raspberry Pi.
      */
-    public static void warmUp(GraphHopperStorage ghStorage, int iterations) {
+    public static void warmUp(GraphHopper graphHopper, int iterations) {
+        GraphHopperStorage ghStorage = graphHopper.getGraphHopperStorage();
         if (ghStorage == null)
             throw new IllegalStateException("Do a successful call to load or importOrLoad before routing");
         if (ghStorage.isClosed())
             throw new IllegalStateException("You need to create a new GraphHopper instance as it is already closed");
-        if(ghStorage.isCHPossible() == false)
-            throw new IllegalStateException("Cannot warm up non CH network");
 
-        BreadthFirstSearch bfs = new BreadthFirstSearch();
-        final Random rand = new Random();
+        if (ghStorage.isCHPossible())
+            warmUpCHSubNetwork(graphHopper, iterations);
+        else
+            warmUpNonCHSubNetwork(graphHopper, iterations);
+    }
+
+    private static void warmUpCHSubNetwork(GraphHopper graphHopper, int iterations) {
+        GraphHopperStorage ghStorage = graphHopper.getGraphHopperStorage();
+
+        for (int i = 0; i < iterations; i++) {
+            int startNode = ThreadLocalRandom.current().nextInt(0, graphHopper.getMaxVisitedNodes() + 1);
+            int endNode = ThreadLocalRandom.current().nextInt(0, graphHopper.getMaxVisitedNodes() + 1);
+
+            double fromLatitude = ghStorage.getNodeAccess().getLatitude(startNode);
+            double fromLongitude = ghStorage.getNodeAccess().getLongitude(startNode);
+            double toLatitude = ghStorage.getNodeAccess().getLatitude(endNode);
+            double toLongitude = ghStorage.getNodeAccess().getLongitude(endNode);
 
+            GHRequest request = new GHRequest(fromLatitude, fromLongitude, toLatitude, toLongitude);
+            graphHopper.route(request);
+        }
+    }
+
+    private static void warmUpNonCHSubNetwork(GraphHopper graphHopper, int iterations) {
+        GraphHopperStorage ghStorage = graphHopper.getGraphHopperStorage();
+
+        BreadthFirstSearch bfs = new BreadthFirstSearch();
         for (int i = 0; i < iterations; i++) {
-            int startNode = rand.nextInt(ghStorage.getBaseGraph().getNodes() + 1);
+            int startNode = ThreadLocalRandom.current().nextInt(0, graphHopper.getMaxVisitedNodes() + 1);
             bfs.start(ghStorage.getBaseGraph().createEdgeExplorer(), startNode);
```

Comment:
The previous nextInt code is correct. The BFS needs to stop after graphHopper.getMaxVisitedNodes() has reached i.e. something like:

```
// it is save to create this once and reuse for others as we have no different states or threads involved:
EdgeExplorer explorer = ghStorage.getBaseGraph().createEdgeExplorer()
for loop {
  // fresh instance necessary to reset counter every time
  new BreadthFirstSearch() {
     int counter = 0;
     public boolean goFurther(int nodeId) {
        counter++;
        return counter < maxVisitedNodes;
     }
  }.start(edgeExplorer, startNode)
}
```

This is not entirely correct as we can visit certain nodes multiple times but we would have to call `bitset.getCardinality` on every call which is too expensive and so not necessary for this 'warmup' purpose.